platform/darwin/src/collator.mm

```diff
@@ -0,0 +1,64 @@
+#include <mbgl/style/expression/collator.hpp>
+
+#include <sstream>
+
+#import <Foundation/Foundation.h>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+class Collator::Impl {
+public:
+    Impl(bool caseSensitive, bool diacriticSensitive, optional<std::string> locale_)
+    : options((caseSensitive ? 0 : NSCaseInsensitiveSearch) |
+              (diacriticSensitive ? 0 : NSDiacriticInsensitiveSearch))
+    , locale(locale_ ?
+                [[NSLocale alloc] initWithLocaleIdentifier:@((*locale_).c_str())] :
+                [NSLocale currentLocale])
```

Comment:
Although `localeIdentifier` is BCP&nbsp;47 compliant when NSLocale is created with a BCP&nbsp;47 locale identifier, `+[NSLocale currentLocale]` places an underscore before the region code, for backwards compatibility. So the one modification we need to make is to replace underscores with hyphens. Otherwise we’re in good shape on iOS and macOS.

Reply:
You mean in `resolvedLocale` we should do the `_`/`-` replacement? I guess since `_` is not a valid part of any BCP 47 tag, that's a relatively safe replacement to do.

Reply:
Added the replacement -- experimentally, it looks like the identifier passed into `initWithLocaleIdentifier` gets round-tripped intact whether it uses the underscore or not. Not sure what pathway `currentLocale` goes down.